neuros/main.py

```diff
@@ -0,0 +1,173 @@
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class WindowConfig:
+    """Configuration for data windowing"""
+    window_ms: float  # Window size in milliseconds
+    overlap_ms: float = 0.0  # Overlap between windows
+
+    def to_samples(self, sample_rate: int) -> tuple[int, int]:
+        """Convert time-based config to sample counts"""
+        window_samples = int(sample_rate * self.window_ms / 1000)
+        overlap_samples = int(sample_rate * self.overlap_ms / 1000)
+        if overlap_samples >= window_samples:
+            raise ValueError("Overlap must be less than window size")
+        return window_samples, overlap_samples
+
+
+@contextmanager
+def board_stream(board_id: int = BoardIds.SYNTHETIC_BOARD,
+                 params: Optional[BrainFlowInputParams] = None) -> BoardShim:
+    """
+    Create and manage a board connection.
+
+    Args:
+        board_id: BrainFlow board identifier
+        params: Optional board parameters
+
+    Yields:
+        Connected BoardShim instance
+
+    Raises:
+        BrainFlowError: If board initialization or streaming fails
+    """
+    board = BoardShim(board_id, params or BrainFlowInputParams())
+    try:
+        board.prepare_session()
+        board.start_stream()
+        yield board
+    except Exception as e:
+        logger.error(f"Error during board operation: {str(e)}")
+        raise
+    finally:
+        try:
+            if board.is_prepared():
+                board.stop_stream()
+                board.release_session()
+        except Exception as e:
+            logger.error(f"Error during board cleanup: {str(e)}")
+            # Don't re-raise during cleanup
+
+
+def stream_windows(board: BoardShim, config: WindowConfig) -> Iterator[np.ndarray]:
+    """
+    Generate windows of EEG data from board stream.
+
+    This generator yields windows of data continuously from the board stream.
+    Each window contains data from all channels (both EEG and non-EEG).
+    Windows can overlap based on configuration, useful for frequency analysis.
+
+    Args:
+        board: Connected and streaming BoardShim instance
+        config: Window configuration for size and overlap
+
+    Yields:
+        Windows of data as numpy arrays with shape (channels, samples)
+        where channels includes all board channels (both EEG and non-EEG)
+        and samples is determined by the window_ms configuration
+
+    Note:
+        The first few windows may be delayed as the board begins
+        collecting data. The generator will continue yielding
+        indefinitely until iteration is stopped.
+    """
+    # Convert ms-based config to samples
+    sample_rate = board.get_sampling_rate(board_id=board.board_id)
+    window_samples, overlap_samples = config.to_samples(sample_rate)
+
+    # Buffer for overlap samples
+    buffer = np.array([])
+
+    # Placeholder for new data
+    new_data = np.array([])
+
+    try:
+        while True:
```

Comment:
Should the loop be caller-side?

Reply:
Opinion: That doesn't change anything functionally, and the caller doesn't really gain any advantage, just an extra job. The current way also makes it visually clearer that we are necessarily working with a stream of data.